src/main/java/io/vertx/core/shareddata/AsyncMapStream.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2011-2017 The original author or authors
+ * ------------------------------------------------------
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Apache License v2.0 which accompanies this distribution.
+ *
+ *     The Eclipse Public License is available at
+ *     http://www.eclipse.org/legal/epl-v10.html
+ *
+ *     The Apache License v2.0 is available at
+ *     http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.vertx.core.shareddata;
+
+import io.vertx.codegen.annotations.VertxGen;
+import io.vertx.core.Closeable;
+import io.vertx.core.Handler;
+import io.vertx.core.streams.ReadStream;
+
+/**
+ * An {@link AsyncMap} keys, values, or entries {@link ReadStream stream}.
+ * <p>
+ * The stream will be automatically closed if it fails or ends.
+ * Otherwise you must invoke {@link #close(Handler)} after usage to avoid leaking resources.
+ *
+ * @author Thomas Segismont
+ */
+@VertxGen
+public interface AsyncMapStream<T> extends ReadStream<T>, Closeable {
```

Comment:
This should rather be named `CollectionStream` or something in that spirit I think as it is used for a collection of key/value or entries from the `AsyncMap`.

Reply:
You're right about the usage (it's explained in the Javadoc). To be honest I also find that `AsyncMapStream` is not self-explanatory. But neither does `CollectionStream` in my opinion. If you find anything better let me know. Otherwise I'll update this one and the cluster managers PRs

Reply:
my point is that this stream could be used for other purpose than `AsyncMapStream` and if we name it `AsyncMapStream` it will feel awkward if we reuse it.

Reply:
`DataStream` might be a good candidate

Reply:
Ha I see. Then how about `CloseableReadstream` ? Isn't a stream always a data stream?

Reply:
Actually I've given thoughts about it and we should instead use plain `ReadStream` and use the `handler(Handler<T>)` to control this like we do in other cases:

- when the handler is set to a value : start to stream the data from the beginning
- when the handler is set to `null` : release the resource

this is how other streams work and how it will work best with RxJava

Reply:
I believe we shall add a `close` operation on `ReadStream` in the future to improve this contract (after 3.5.0).

Reply:
So IIUC, setting a null handler would close the stream? Can you elaborate on how this will work with RxJava?

Reply:
>Setting a null handler would close the stream and setting a non null handler would start it (lazy start).

OK, thanks for clarifying.

>When a ReadStream is adapted to an Observable its handlers are set at subscription and unset at subscription cancellation. You can look at the implementation of MessageConsumer that works this way.

So that will work because the stream implementations in cluster managers don't create the resources until you set the data handler.

Alright, I'll switch back to a plain read stream, update the Javadoc and cluster manager implementations.

Reply:
Done